### PR TITLE
Improves PandasConverter

### PR DESCRIPTION
**Pull request by AlexCatarino**
In the previous implementation, each one of bars.Select statement causes an extra enumeration of the bars enumerable. To make matters worse, this is all invoked 4*N times, where N is the number of data points.

Another bottleneck was related to the way we concatenate pandas.DataFrames: we would add a new data frame to an existing data frame individually. Now, we collect all data frames and concatenate them in one operation.

With this new implementation, we have accomplshed a reduction of 90% of memory usage in a history request of 1000 symbols.

### CODE REVIEW

File: Common/Python/PandasConverter.cs [hunk 1]
```diff
@@ -61,149 +61,271 @@ public PandasConverter(PyObject pandas = null)
...
-                        GetDataFrame(data.Get<TradeBar>(symbol))
-                    };
-
-                    dataFrame = _pandas.concat(new PyList(items));
+                    sliceDataDict[baseData.Symbol].Add(baseData);
                 }
+            }
 
-                return dataFrame;
+            var maxLevels = sliceDataDict.Max(x => x.Value.Levels);
```
Comment (mchandschuh) [inline Common/Python/PandasConverter.cs]: Can we compute this during the iteration above?

File: Common/Python/PandasConverter.cs [hunk 2]
```diff
@@ -61,149 +61,271 @@ public PandasConverter(PyObject pandas = null)
...
-            
-            using (Py.GIL())
-            {
-                var dataFrame = _pandas.DataFrame();
-
-                foreach (var symbol in symbols)
+                foreach (var baseData in slice.Values)
                 {
-                    if (symbol == underlying)
+                    if (!sliceDataDict.ContainsKey(baseData.Symbol))
```
Comment (mchandschuh) [inline Common/Python/PandasConverter.cs]: If we use TryGetValue then we won't need the additional dictionary access below

File: Common/Python/PandasConverter.cs [hunk 3]
```diff
@@ -61,149 +61,271 @@ public PandasConverter(PyObject pandas = null)
...
-                pyDict.SetItem("high", _pandas.Series(data.Select(x => (double)x.High).ToList(), index));
-                pyDict.SetItem("close", _pandas.Series(data.Select(x => (double)x.Close).ToList(), index));
 
-                if (typeof(T) == typeof(TradeBar))
+                foreach (var kvp in data)
                 {
-                    Func<IBaseDataBar, double> getVolume = x => { var bar = x as TradeBar; return (double)bar.Volume; };
-                    pyDict.SetItem("volume", _pandas.Series(data.Select(x => getVolume(x)).ToList(), index));
+                    var index = kvp.Value.Select(x => x.EndTime).ToList();
+                    var values = kvp.Value.Select(x => (double)x.Value).ToList();
```
Comment (mchandschuh) [inline Common/Python/PandasConverter.cs]: We could build both of these lists in a single iteration of kvp.Value

File: Common/Python/PandasConverter.cs [hunk 4]
```diff
@@ -61,149 +61,271 @@ public PandasConverter(PyObject pandas = null)
...
+    }
+
+    /// <summary>
+    /// Organizes a list of data to create pandas.DataFrames 
+    /// </summary>
+    public class PandasData
+    {
+        private Symbol _symbol;
+        private List<DateTime> _timeIndex;
+        private Dictionary<string, List<double>> _series;
```
Comment (mchandschuh) [inline Common/Python/PandasConverter.cs]: Should the be read only?

File: Common/Python/PandasConverter.cs [hunk 5]
```diff
@@ -61,149 +61,271 @@ public PandasConverter(PyObject pandas = null)
...
+        /// Implied levels of a multi index pandas.Series (depends on the security type)
+        /// </summary>
+        public int Levels
+        {
+            get
+            {
+                var levels = 2;
+                if (_symbol.SecurityType == SecurityType.Future) levels += 1;
+                if (_symbol.SecurityType == SecurityType.Option) levels += 3;
+                return levels;
```
Comment (mchandschuh) [inline Common/Python/PandasConverter.cs]: Why not calculate this once in constructor?

File: Common/Python/PandasConverter.cs [hunk 6]
```diff
@@ -61,149 +61,271 @@ public PandasConverter(PyObject pandas = null)
...
+            _timeIndex = new List<DateTime>();
+        }
+
+        /// <summary>
+        /// Adds an object to the end of the lists
+        /// </summary>
+        /// <param name="baseData"><see cref="IBaseData"/> object that contains information to be saved in an instance of <see cref="PandasData"/></param>
+        public void Add(IBaseData baseData)
+        {
+            _timeIndex.Add(baseData.Time);
```
Comment (mchandschuh) [inline Common/Python/PandasConverter.cs]: Should we be using time or end time? Philosophical question really. For daily data people would expect time, the date of the data, vs other data end time makes sense. I want the data emitted at 9:42:00 vs saying I want the data that started at 9:39:59

Comment (mchandschuh) [reply Common/Python/PandasConverter.cs]: Perhaps we need a test of this pandas converter that puts data on and reads it back out and confirms the data is all the same and accessible in the same way

File: Common/Python/PandasConverter.cs [hunk 7]
```diff
@@ -61,149 +61,271 @@ public PandasConverter(PyObject pandas = null)
...
+                }
             }
 
-            return _pandas.MultiIndex.from_tuples(tuples.ToArray(), names: names.Split(','));
+            var data = baseData as DynamicData;
+            if (data != null)
+            {
+                foreach (var kvp in data.GetStorageDictionary())
+                {
+                    _series[kvp.Key].Add(Convert.ToDouble(kvp.Value));
```
Comment (mchandschuh) [inline Common/Python/PandasConverter.cs]: I think we want the value property also for custom data types, such as Bitcoin

**Comment by AlexCatarino** [comment]
The proposed changes were made. Thank you for the input.
Concerning the performance, it got about 40-50% faster. The 1000-equity history request before the change took 45-50 seconds and it now takes 24-26 seconds.

File: Common/Python/PandasConverter.cs (hunk 7 above)
Comment (AlexCatarino) [reply Common/Python/PandasConverter.cs]: That was included in the line below. :)

File: Common/Python/PandasConverter.cs [hunk 8]
```diff
@@ -189,7 +193,7 @@ public PandasData(IBaseData baseData)
         /// <param name="baseData"><see cref="IBaseData"/> object that contains information to be saved in an instance of <see cref="PandasData"/></param>
         public void Add(IBaseData baseData)
         {
-            _timeIndex.Add(baseData.Time);
+            _timeIndex.Add(baseData.EndTime);
```
Comment (mchandschuh) [inline Common/Python/PandasConverter.cs]: I was hoping for a little discussion around this change, @jaredbroad -- any comments?

One issue I see is with daily data. To access data *for* to 30th of October I would need to enter a date of the 31st

Comment (AlexCatarino) [reply Common/Python/PandasConverter.cs]: I think we should go for EndTime. If the users makes a History request to warm up and indicator, he should use EndTime in Indicator.Update method. If we give him Time. he would need to discount the Period.

Comment (mchandschuh) [reply Common/Python/PandasConverter.cs]: That's a pretty good point, also, I think it's rare people are accessing the data frame by date/time

**Pull request merged**